File: python/src/rivu_server_sdk/snapshot_store.py
from __future__ import annotations

import json
import sqlite3
import time
from dataclasses import dataclass
from pathlib import Path
from typing import Any


@dataclass(frozen=True)
class Snapshot:
    thread_id: str
    seq: int
    shared_state: dict[str, Any]
    created_at_ms: int


class SqliteSnapshotStore:
    def __init__(self, path: str | Path) -> None:
        self._path = str(path)
        self._init_db()

    @property
    def path(self) -> str:
        return self._path

    def _connect(self) -> sqlite3.Connection:
        conn = sqlite3.connect(self._path)
        conn.execute("PRAGMA journal_mode=WAL;")
        conn.execute("PRAGMA synchronous=NORMAL;")
        return conn

    def _init_db(self) -> None:
        Path(self._path).parent.mkdir(parents=True, exist_ok=True)
        with self._connect() as conn:
            conn.execute(
                """
                CREATE TABLE IF NOT EXISTS rivu_snapshots (
                  thread_id TEXT NOT NULL,
                  seq INTEGER NOT NULL,
                  snapshot_json TEXT NOT NULL,
                  created_at_ms INTEGER NOT NULL,
                  PRIMARY KEY(thread_id, seq)
                )
                """,
            )
            conn.execute(
                "CREATE INDEX IF NOT EXISTS idx_rivu_snapshots_thread_seq ON rivu_snapshots(thread_id, seq)",
            )

    def put(self, *, thread_id: str, seq: int, shared_state: dict[str, Any]) -> None:
        if not thread_id.strip():
            raise ValueError("thread_id must be non-empty")
        if not isinstance(seq, int) or seq <= 0:
            raise ValueError("seq must be a positive int")
        if not isinstance(shared_state, dict):
            raise ValueError("shared_state must be a dict")

        created_at_ms = int(time.time() * 1000)
        payload = json.dumps(shared_state, ensure_ascii=False, separators=(",", ":"))
        with self._connect() as conn:
            conn.execute(
                "INSERT OR REPLACE INTO rivu_snapshots(thread_id, seq, snapshot_json, created_at_ms) VALUES (?,?,?,?)",
                (thread_id, seq, payload, created_at_ms),
            )

    def get_latest(self, *, thread_id: str) -> Snapshot | None:
        if not thread_id.strip():
            raise ValueError("thread_id must be non-empty")
        with self._connect() as conn:
            row = conn.execute(
                "SELECT seq, snapshot_json, created_at_ms FROM rivu_snapshots WHERE thread_id=? ORDER BY seq DESC LIMIT 1",
                (thread_id,),
            ).fetchone()
        if not row:
            return None
        seq, snapshot_json, created_at_ms = row
        shared_state = json.loads(snapshot_json)
        if not isinstance(shared_state, dict):
            raise ValueError("stored snapshot is not an object")
        return Snapshot(thread_id=thread_id, seq=int(seq), shared_state=shared_state, created_at_ms=int(created_at_ms))

File: crates/rivu-server-sdk/src/snapshot_store.rs
use std::fs;
use std::path::{Path, PathBuf};
use std::time::{SystemTime, UNIX_EPOCH};

use rusqlite::{params, Connection, OptionalExtension};
use serde::{Deserialize, Serialize};
use serde_json::Value;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Snapshot {
    pub thread_id: String,
    pub seq: u64,
    pub shared_state: Value,
    pub created_at_ms: i64,
}

#[derive(thiserror::Error, Debug)]
pub enum SnapshotStoreError {
    #[error("thread_id must be non-empty")]
    EmptyThreadId,
    #[error("seq must be a positive integer")]
    InvalidSeq,
    #[error("snapshot must be a json object")]
    SnapshotNotObject,
    #[error("sqlite error: {0}")]
    Sqlite(#[from] rusqlite::Error),
    #[error("json error: {0}")]
    Json(#[from] serde_json::Error),
    #[error("io error: {0}")]
    Io(#[from] std::io::Error),
}

#[derive(Debug, Clone)]
pub struct SqliteSnapshotStore {
    path: PathBuf,
}

impl SqliteSnapshotStore {
    pub fn new(path: impl Into<PathBuf>) -> Result<Self, SnapshotStoreError> {
        let path = path.into();
        if let Some(parent) = path.parent() {
            if !parent.as_os_str().is_empty() {
                fs::create_dir_all(parent)?;
            }
        }
        let store = Self { path };
        store.init_db()?;
        Ok(store)
    }

    pub fn path(&self) -> &Path {
        &self.path
    }

    fn open(&self) -> Result<Connection, SnapshotStoreError> {
        let conn = Connection::open(&self.path)?;
        conn.pragma_update(None, "journal_mode", "WAL")?;
        conn.pragma_update(None, "synchronous", "NORMAL")?;
        Ok(conn)
    }

    fn init_db(&self) -> Result<(), SnapshotStoreError> {
        let conn = self.open()?;
        conn.execute_batch(
            r#"
            CREATE TABLE IF NOT EXISTS rivu_snapshots (
              thread_id TEXT NOT NULL,
              seq INTEGER NOT NULL,
              snapshot_json TEXT NOT NULL,
              created_at_ms INTEGER NOT NULL,
              PRIMARY KEY(thread_id, seq)
            );
            CREATE INDEX IF NOT EXISTS idx_rivu_snapshots_thread_seq ON rivu_snapshots(thread_id, seq);
            "#,
        )?;
        Ok(())
    }

    pub fn put(&self, thread_id: &str, seq: u64, shared_state: &Value) -> Result<(), SnapshotStoreError> {
        if thread_id.trim().is_empty() {
            return Err(SnapshotStoreError::EmptyThreadId);
        }
        if seq == 0 {
            return Err(SnapshotStoreError::InvalidSeq);
        }
        if !shared_state.is_object() {
            return Err(SnapshotStoreError::SnapshotNotObject);
        }

        let created_at_ms = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap_or_default()
            .as_millis() as i64;
        let payload = serde_json::to_string(shared_state)?;

        let conn = self.open()?;
        conn.execute(
            "INSERT OR REPLACE INTO rivu_snapshots(thread_id, seq, snapshot_json, created_at_ms) VALUES (?,?,?,?)",
            params![thread_id, seq as i64, payload, created_at_ms],
        )?;
        Ok(())
    }

    pub fn get_latest(&self, thread_id: &str) -> Result<Option<Snapshot>, SnapshotStoreError> {
        if thread_id.trim().is_empty() {
            return Err(SnapshotStoreError::EmptyThreadId);
        }
        let conn = self.open()?;
        let mut stmt = conn.prepare(
            "SELECT seq, snapshot_json, created_at_ms FROM rivu_snapshots WHERE thread_id=? ORDER BY seq DESC LIMIT 1",
        )?;
        let row = stmt
            .query_row(params![thread_id], |row| {
                let seq: i64 = row.get(0)?;
                let snapshot_json: String = row.get(1)?;
                let created_at_ms: i64 = row.get(2)?;
                Ok((seq, snapshot_json, created_at_ms))
            })
            .optional()?;

        match row {
            None => Ok(None),
            Some((seq, json, created_at_ms)) => {
                let shared_state: Value = serde_json::from_str(&json)?;
                if !shared_state.is_object() {
                    return Err(SnapshotStoreError::SnapshotNotObject);
                }
                Ok(Some(Snapshot {
                    thread_id: thread_id.to_string(),
                    seq: seq as u64,
                    shared_state,
                    created_at_ms,
                }))
            }
        }
    }
}
